### DISCUSSION TITLE
[BUG] CSOM service exception in Set-PnPPageTextPart on Azure Functions

### DISCUSSION THREAD

**Discussion by Rabomax** [pnp/powershell#3272] [General]
### Reporting an Issue or Missing Feature
I'm coding a PowerShell script within an Azure Function App that translates SharePoint pages with the help of the PnP.PowerShell module and Azure Translator

### Expected behavior 
SharePoint page should be translated after calling _**Set-PnPPageTextPart** -Page $newPage -InstanceId $instanceId -Text $translatedControlText_

### Actual behavior 
Getting an CSOM service exception when calling _**Set-PnPPageTextPart**_

Exception             : 
    Type        : System.Management.Automation.PSInvalidOperationException
    ErrorRecord : 
        Exception             : 
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : CSOM service exception
            HResult : -2146233087
        CategoryInfo          : InvalidOperation: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : InvalidOperation
    TargetSite  : 
        Name          : ProcessRecord
        DeclaringType : PnP.PowerShell.Commands.Base.PnPConnectedCmdlet
        MemberType    : Method
        Module        : PnP.PowerShell.dll
    Message     : CSOM service exception
    Source      : PnP.PowerShell
    HResult     : -2146233079
    StackTrace  : 
   at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPConnectedCmdlet.cs:line 98
   at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPSharePointCmdlet.cs:line 120
   at System.Management.Automation.Cmdlet.DoProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()
CategoryInfo          : InvalidOperation: (:) [Set-PnPPageTextPart], PSInvalidOperationException
FullyQualifiedErrorId : InvalidOperation,PnP.PowerShell.Commands.Pages.SetClientSideText
InvocationInfo        : 
    MyCommand        : Set-PnPPageTextPart
    ScriptLineNumber : 116
    OffsetInLine     : 9
    HistoryId        : 1
    ScriptName       : C:\home\site\wwwroot\OnNewSite\run.ps1
    Line             : Set-PnPPageTextPart -Page $newPage -InstanceId $textControl.InstanceId -Text $translatedControlText
                       
    PositionMessage  : At C:\home\site\wwwroot\OnNewSite\run.ps1:116 char:9
                       +         Set-PnPPageTextPart -Page $newPage -InstanceId $textControl.I …
                       +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\home\site\wwwroot\OnNewSite
    PSCommandPath    : C:\home\site\wwwroot\OnNewSite\run.ps1
    InvocationName   : Set-PnPPageTextPart
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, C:\home\site\wwwroot\OnNewSite\run.ps1: line 116

Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException : Result: ERROR: CSOM service exception

Exception             : 
    Type        : System.Management.Automation.PSInvalidOperationException
    ErrorRecord : 
        Exception             : 
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : CSOM service exception
            HResult : -2146233087
        CategoryInfo          : InvalidOperation: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : InvalidOperation
    TargetSite  : 
        Name          : ProcessRecord
        DeclaringType : PnP.PowerShell.Commands.Base.PnPConnectedCmdlet
        MemberType    : Method
        Module        : PnP.PowerShell.dll
    Message     : CSOM service exception
    Source      : PnP.PowerShell
    HResult     : -2146233079
    StackTrace  : 
   at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPConnectedCmdlet.cs:line 98
   at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPSharePointCmdlet.cs:line 120
   at System.Management.Automation.Cmdlet.DoProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()
CategoryInfo          : InvalidOperation: (:) [Set-PnPPageTextPart], PSInvalidOperationException
FullyQualifiedErrorId : InvalidOperation,PnP.PowerShell.Commands.Pages.SetClientSideText
InvocationInfo        : 
    MyCommand        : Set-PnPPageTextPart
    ScriptLineNumber : 116
    OffsetInLine     : 9
    HistoryId        : 1
    ScriptName       : C:\home\site\wwwroot\OnNewSite\run.ps1
    Line             : Set-PnPPageTextPart -Page $newPage -InstanceId $textControl.InstanceId -Text $translatedControlText
                       
    PositionMessage  : At C:\home\site\wwwroot\OnNewSite\run.ps1:116 char:9
                       +         Set-PnPPageTextPart -Page $newPage -InstanceId $textControl.I …
                       +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\home\site\wwwroot\OnNewSite
    PSCommandPath    : C:\home\site\wwwroot\OnNewSite\run.ps1
    InvocationName   : Set-PnPPageTextPart
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, C:\home\site\wwwroot\OnNewSite\run.ps1: line 116

Exception: CSOM service exception
Stack:    at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPConnectedCmdlet.cs:line 98
   at PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPSharePointCmdlet.cs:line 120
   at System.Management.Automation.Cmdlet.DoProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()

### Steps to reproduce behavior

1. Deploy Azure Function with managed identity and certificate and accessing app registration with application permission _SharePoint Online.Sites.Selected_ and _Graph.Sites.Selected_
2. Set permissions for Sites.Selected with **Grant-PnPAzureADAppSitePermission**
3. Write code as following

try {
      Connect-PnPOnline -Url $siteURL -Tenant $ten -ClientId $app -CertificateBase64Encoded $certificateBase64Encoded
}
catch {
      Write-Host ($Error[0].Exception)
}
  
$newPage = Get-PnPClientSidePage "$targetLanguage/$pageTitle"
$textControls = $newPage.Controls | Where-Object {$_.Type.Name -eq "PageText"}
  
Write-Host "After retrieved controls..."
  
foreach ($textControl in $textControls) {
      $comp = Get-PnPPageComponent -Page $newPage -InstanceId $textControl.InstanceId
      $translatedControlText = Start-Translation -text $comp.Text -language $targetLanguage
      Write-Host $translatedControlText
      Set-PnPPageTextPart -Page $newPage -InstanceId $textControl.InstanceId -Text $translatedControlText
}

### What is the version of the Cmdlet module you are running?

- PnP.PowerShell = 2.1.1
- Az.* = 9.*
- Azure Function Runtime = 4.19.0.20414
- PowerShell Runtime version = 7.2 (with 7.0 it also doesn't work)
- Azure online code editor

### Which operating system/environment are you running PnP PowerShell on?
- [ X ] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ X ] Azure Functions
- [ ] Other : please specify

### Additional context
Everything works fine, from establishing the connection with Connect-PnPOnline to retrieving the text webparts of the given page but it always fails on **Set-PnPPageTextPart**. I have also followed #2795 and updated PnP.PowerShell from 1.12 to 2.1.1 but it also did not work.

**Comment by gautamdsheth** [comment]
Are you using or loading any other modules in the runtime like Az or Graph or some other ? If so, can you please comment them out and try ?

**Comment by Rabomax** [comment]
> Are you using or loading any other modules in the runtime like Az or Graph or some other ? If so, can you please comment them out and try ?

Hi Gautam,

thanks for you reply.

requirements.psd1 ==>

`@{
    'Az' = '9.*'
    'PnP.PowerShell' = '2.1.1'
}`

profile.ps1 ==>

Import-Module Az.Accounts #-RequiredVersion '2.12.2'
Import-Module PnP.PowerShell #-RequiredVersion '2.1.1'

if ($env:MSI_SECRET) {
    Disable-AzContextAutosave -Scope Process | Out-Null
    Connect-AzAccount -Identity
}

Kind regards,
Markus

ad unloading other modules) un-commenting **Import-Module Az,Accounts** from _profile.ps1_ has no effect.

**Comment by Rabomax** [comment]
Hello @gautamdsheth are there any new experience values regarding this bug?

**Comment by Rabomax** [comment]
Hello @gautamdsheth unfortunately it still does not work. Is there any update on it?

Kind regards,
Markus

**Comment by gautamdsheth** [comment]
@Rabomax - unfortunately can't reproduce this issue at my end.

<img width="854" alt="image" src="https://github.com/pnp/powershell/assets/8832018/08cfa3ad-e9f5-45e9-8843-57b965bcfeb4">

Looks like some other module or the translation function might be causing the issue.
Please try with hardcoded text in `$translatedText` variable and try.

This is my requirements.psd1 file, as you can see, I only have PnP PowerShell loaded.

<img width="588" alt="image" src="https://github.com/pnp/powershell/assets/8832018/51a671d9-6b4d-4984-9d97-01450e633cbe">

My profile.ps1 is totally commented out:

<img width="662" alt="image" src="https://github.com/pnp/powershell/assets/8832018/c9054f13-4110-48f3-8bce-88bb253d9ec4">


Will move it to a discussion now.
